sys/cddl/dev/dtrace/dtrace_ioctl_compat.c

```diff
@@ -0,0 +1,260 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+ * or http://www.opensolaris.org/os/licensing.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ *
+ * $FreeBSD$
+ *
+ */
+
+static void * __capability
+dtrace_make_buffdesc_cap(caddr_t addr)
+{
+#ifdef COMPAT_FREEBSD64
+	if (!SV_CURPROC_FLAG(SV_CHERI))
+	    return __USER_CAP_OBJ(*(dtrace_bufdesc64_t **)addr);
+#endif
+	return *(dtrace_bufdesc_t * __capability *)addr;
+}
+
+static int
+dtrace_copyin_buffdesc(
+    dtrace_bufdesc_t * __capability uaddr, dtrace_bufdesc_t *bufdesc)
+{
+#ifdef COMPAT_FREEBSD64
+	if (!SV_CURPROC_FLAG(SV_CHERI)) {
+		dtrace_bufdesc64_t bufdesc64;
+		int retval = copyin(uaddr, &bufdesc64, sizeof (bufdesc64));
+		bufdesc->dtbd_size = bufdesc64.dtbd_size;
+		bufdesc->dtbd_cpu = bufdesc64.dtbd_cpu;
+		bufdesc->dtbd_errors = bufdesc64.dtbd_errors;
+		bufdesc->dtbd_drops = bufdesc64.dtbd_drops;
+		bufdesc->dtbd_data = __USER_CAP_STR(bufdesc64.dtbd_data);
+		bufdesc->dtbd_oldest = bufdesc64.dtbd_oldest;
+		bufdesc->dtbd_timestamp = bufdesc64.dtbd_timestamp;
+		return retval;
+	}
+#endif
+	return copyincap(uaddr, bufdesc, sizeof (dtrace_bufdesc_t));
+}
+
+static int
+dtrace_copyout_buffdesc(dtrace_bufdesc_t *bufdesc,
+    dtrace_bufdesc_t * __capability uaddr)
+{
+#ifdef COMPAT_FREEBSD64
+	if (!SV_CURPROC_FLAG(SV_CHERI)) {
+		dtrace_bufdesc64_t bufdesc64;
+		bufdesc64.dtbd_size = bufdesc->dtbd_size;
+		bufdesc64.dtbd_cpu = bufdesc->dtbd_cpu;
+		bufdesc64.dtbd_errors = bufdesc->dtbd_errors;
+		bufdesc64.dtbd_drops = bufdesc->dtbd_drops;
+		bufdesc64.dtbd_data = (__cheri_addr uint64_t)bufdesc->dtbd_data;
+		bufdesc64.dtbd_oldest = bufdesc->dtbd_oldest;
+		bufdesc64.dtbd_timestamp = bufdesc->dtbd_timestamp;
+		return copyout(&bufdesc64, uaddr, sizeof (dtrace_bufdesc64_t));
+	}
+#endif
+	return copyoutcap(bufdesc, uaddr, sizeof (dtrace_bufdesc_t));
+}
+
+static void
+dtrace_bcopy_recdesc(dtrace_recdesc_t *recdesc, void *dest)
```

Comment:
I think I'd prefer these be named `_copy_` and not allude to a semi-deprecated function.  Also, the source should be `const`.

Reply:
fixed